README
= RExchange

*July 28th 2008 - This is a fork of the stagnant rubyforge project. For historical see http://rubyforge.org/projects/rexchange/*

RExchange is a pure ruby wrapper for the Microsoft Exchange Server WebDAV API

== Things you should know

* Requires Ruby 1.8.4 (or later, for the extended WebDAV support in the net/http library)
* RExchange is cross-platform compatible, being written in pure Ruby
* Kiwi fruits are packed with vitamins

== Why should you use RExchange

* It makes interacting with Exchange simple
* It was written for a real application, and does real work reliably day in and day out

== Example

	# We pass our uri (pointing directly to a mailbox), a username, and a password to RExchange::open
	# to create a RExchange::Session. Notice that we escape the "\" in our username.
	RExchange::open('https://example.com/exchange/admin/', 'mydomain\\admin', 'secret') do |mailbox|
  
		# The block parameter ("mailbox" in this case) is actually the Session itself.
		# You can refer to folders by chaining them as method calls. "inbox" in this case
		# isn't a defined method for Session, but part of the DSL to refer to folder names.
		# Each folder name returns a RExchange::Folder. The folder is Enumerable, allowing
		# iteration over the items in the folder, depending on the DAV:content-class of
		# the folder.
		mailbox.inbox.each do |message|
    
			# The "message" block parameter is a RExchange::Message object. You have access to
			# several attributes of the message, including: href, from, to, message_id, date,
			# importance, has_attachment? and body.
			p message.subject
    
			# The RExchange::Message#move_to method moves the message to another folder, in this
			# case, an "archive" folder off of the inbox.
			message.move_to mailbox.inbox.archive
		end
  
		# Folder names are "normalized", replacing dashes and spaces with underscores,
		# squeezing out multiple underscores in a row, and downcasing the whole thing.
		# So a folder name such as "My Very-long Folder Name" would look like:
		mailbox.my_very_long_folder_name
  
		# The Enumerable mixin gives us an Folder#entries method in case we just want to
		# return an array.
		messages = mailbox.inbox.entries
	
		# Other methods, like Enumerable#map and Enumerable#sort_by let us manipulate
		# the Folder in fun ways:
		puts mailbox.contacts.sort_by { |contact| contact.first_name }.map do |contact|
			<<-CONTACT
				Name:    #{contact.first_name} #{contact.last_name}
				Email:   #{contact.email}
				Phone:   #{contact.phone}
			CONTACT
		end
	
	end

	# We can access the Global Address Book by accessing the "/public" folder:
	RExchange::open('https://example.com/public', 'mydomain\\somebody', 'secret') do |mailbox|

		puts mailbox.company_contacts.each do |contact|
			<<-CONTACT
				Company: #{contact.company}
				Name:    #{contact.first_name} #{contact.last_name}
				Email:   #{contact.email}
				Phone:   #{contact.phone}
			CONTACT
	 end

	end
	
== Caveats

There are several features missing (simply because we didn't need them yet). Among them:

* The ability to delete messages or folders
* The ability to create folders
* There's no mechanism for sending new messages, or replying or forwarding existing ones
* There are a lot more message attributes we're not retrieving since they weren't useful to us, but they might be to you
* Exporting an email or entire folder tree to offline storage
* And much much more!

If you'd like to see any of these features, or have some ideas of your own you'd like to see implemented don't hesitate to let us know, and if it strikes our fancy maybe you'll get some free programming!

CHANGELOG
-- 0.4.0:  July 28th 2008 
* Forked the stagnant rubyforge project into a new on on github. For historical reference see 
  http://rubyforge.org/projects/rexchange/


-- 0.3.4:
* Updated the README so it jives with the current version
* Removed Folder::join since it's only used by Folder#to_s, and just used a simpler
	solution in Folder#to_s instead

-- 0.3.3:
* Folder objects now default to using the Message class for enumeration 
  if no class is associated with the content_type of the Folder.
  
-- 0.3.2:
* Quick hack to make Message#move_to work

-- 0.3.1:
* Disabled a warning about overwriting GenericItem::CONTENT_CLASS
* Added a "rescue" for Time::parse failures that reverts to the original element text

-- 0.3.0:
* Added support for Appointments
* Options hash for Session creation is gone. The new form is: "RExchange::Session.new(url, username, password)"
* Removed GenericItem#[], GenericItem#method_missing (attributes are explicit now)
* Removed Folder#get_messages, #get_contacts and #messages_in. Folders are now Enumerables using Enumerable#entries, #each, etc
  to retrieve entries according to the folder's content-class. So a 'mailfolder' will retrieve Message instances,
  a 'calendarfolder' will retrieve Appointment instances, and so on.

-- 0.2.0:
* Fixed a bug in Message#move_to when passing folders
* Added support for Contact browsing
* Added GenericItem to abstract Contacts and Messages
* Renamed Folder#messages to Folder#get_messages to avoid naming collisions
* Modified attributes in GenericItem that end with "date" to be Time::parse'd
* Added "r_exchange.rb" for Rails auto-require compatibility
* Now calling String#normalize on attribute names instead of just String#tr('-', '_')

-- 0.1.4:
* Moved Folder::normalize_folder_name to String#normalize, and added support for MixedCase sources.
* Fixed a bug where the normalized folder name was being used for the Folder#to_s result.
* Moved the meat of Folder#messages to Message::find
* Added Message#to_s
* Modified Folder#get_folders to return a hash instead of an array, providing both the original, and normalized names.

-- 0.1.3: Bugfixes
* Fixed several nasty bugs.

-- 0.1.2: Minor documentation/cleanup
* Removed a double-dash in the README that was apparently causing some viewing errors
* Changed Session initialize to a ternary condition
* Added RExchange::Message#has_attachments?
* Added documentation for most methods in RExchange::Message and RExchange::Folder

-- 0.1.1
* Fleshed out the README
* Copied the RAKEFILE from the Rools project

lib/rexchange/folder.rb
require 'net/https'
require 'rexchange/dav_search_request'
require 'rexchange/message'
require 'rexchange/contact'
require 'rexchange/attachment'
require 'rexchange/appointment'
require 'rexchange/note'
require 'rexchange/task'

module RExchange

  class FolderNotFoundError < StandardError
  end
  
  class Folder
   
    attr_reader :credentails, :name
    
    def initialize(credentials, parent, name, content_type)
      @credentials, @parent, @name = credentials, parent, name
      @content_type = CONTENT_TYPES[content_type] || Message
    end
    
    # Used to access subfolders.
    def method_missing(sym, *args)
      if folders.has_key?(sym.to_s)
        folders[sym.to_s]
      else
        raise FolderNotFoundError.new("#{sym} is not a subfolder of #{name}")
      end
    end
    
    include Enumerable
    
    # Iterate through each entry in this folder
    def each
      @content_type::find(@credentials, to_s).each do |item|
        yield item
      end
    end
    
    # Not Implemented!
    def search(conditions = {})
      raise NotImplementedError.new('Bad Touch!')
      @content_type::find(@credentials, to_s, conditions)
    end
    
    # Return an Array of subfolders for this folder
    def folders
      @folders ||= begin
        request_body = <<-eos
      		<?xml version="1.0"?>
  				<D:searchrequest xmlns:D = "DAV:">
  					 <D:sql>
  					 SELECT "DAV:displayname", "DAV:contentclass"
  					 FROM SCOPE('shallow traversal of "#{to_s}"')
  					 WHERE "DAV:ishidden" = false
                         AND "DAV:isfolder" = true
  					 </D:sql>
  				</D:searchrequest>
        eos

        response = DavSearchRequest.execute(@credentials, :body => request_body)

        folders = {}

        # iterate through folders query and add a new Folder
        # object for each, under a normalized name.
        xpath_query = "//a:propstat[a:status/text() = 'HTTP/1.1 200 OK']/a:prop"
        XML::Parser.string(response.body).parse.find(xpath_query).each do |m|
          displayname = m.find_first('a:displayname').content
          contentclass = m.find_first('a:contentclass').content
          folders[displayname.normalize] = Folder.new(@credentials, self, displayname, contentclass.split(':').last.sub(/folder$/, ''))          
        end
        
        folders
      end
    end
    
    # Return the absolute path to this folder (but not the full URI)
    def to_s
      "#@parent/#@name/".squeeze('/')
    end
  end
end


lib/rexchange/message.rb
require 'rexchange/generic_item'
require 'rexchange/dav_enum_atts_request'
require 'rexchange/dav_delete_request'

module RExchange
  class Message < GenericItem
    
    set_folder_type 'mail'
    
    attribute_mappings :from => 'urn:schemas:httpmail:from',
      :to => 'urn:schemas:httpmail:to',
      :message_id => 'urn:schemas:mailheader:message-id',
      :subject => 'urn:schemas:httpmail:subject',
      :recieved_on => 'urn:schemas:httpmail:date',
      :importance => 'urn:schemas:httpmail:importance',
      :has_attachments? => 'urn:schemas:httpmail:hasattachment',
      :body => 'urn:schemas:httpmail:textdescription',
      :html => 'urn:schemas:httpmail:htmldescription'
    
    # Move this message to the specified folder.
    # The folder can be a string such as 'inbox/archive' or a RExchange::Folder.
    # === Example
    #   mailbox.inbox.each do |message|
    #     message.move_to mailbox.inbox.archive
    #   end
    def move_to(folder)
      source = URI.parse(self.href).path
      destination = if folder.is_a?(RExchange::Folder)
        folder.to_s.ensure_ends_with('/') + source.split('/').last
      else
        @session.uri.path.ensure_ends_with('/') + folder.to_s.ensure_ends_with('/') + source.split('/').last
      end

      # $log.debug "move_to: source => \"#{source}\", destination => \"#{destination}\""
      DavMoveRequest.execute(@session, source, destination)
    end
    

    def attachments
      DavEnumAttsRequest.execute(@session, self.href)
    end
    
    def delete
      DavDeleteRequest.execute(@session, self.href)
    end
            
    def to_s
      "To: #{to}, From: #{from}, Subject: #{subject}"
    end
    
  end
end

lib/rexchange/dav_enum_atts_request.rb
require 'rexchange/exchange_request'

module RExchange
  class DavEnumAttsRequest < ExchangeRequest
    METHOD = 'X-MS-ENUMATTS'
    REQUEST_HAS_BODY = false
    RESPONSE_HAS_BODY = true
    
    def self.execute(credentials, source)
      options = {
        :headers => {
          'Depth' => 1,
          'Content-type' => "xml"
        },
        :path => source
      }
      
      response = super(credentials, options)
      items = []
      attachment_xpath = "//a:response"
      properties_xpath = "//a:propstat[a:status/text() = 'HTTP/1.1 200 OK']/a:prop"

      XML::Parser.string(response.body).parse.find(attachment_xpath).each do |m|
        a = Attachment.new(credentials, m.find_first(properties_xpath))
        a.href = m.find_first("a:href").content
        items << a        
      end

      items
    end
  end
end

lib/rexchange/generic_item.rb
require 'ostruct'
require 'rexchange/dav_move_request'
require 'time'
require 'rubygems'
require 'xml'


module RExchange
  
  class Folder
    CONTENT_TYPES = {}
  end
  
  class GenericItem
    include Enumerable

    attr_accessor :attributes

    def initialize(session, dav_property_node)
      @attributes = {}
      @session = session

      dav_property_node.children.each do |element|
        namespaced_name = element.namespace_node.href + element.name
        
        if element.name =~ /date$/i || self.class::ATTRIBUTE_MAPPINGS.find { |k,v| v == namespaced_name && k.to_s =~ /\_(at|on)$/ }
           @attributes[namespaced_name] = DateTime::parse(element.content) rescue element.content
        elsif element.name == 'keywords-utf8'
           @attributes[namespaced_name] = element.to_a.collect! { |x| x.content } rescue nil          
        else
           @attributes[namespaced_name] = element.content
        end
      end
    end

    # Set the default CONTENT_CLASS to the class name, and define a
    # dynamic query method for the derived class.
    def self.inherited(base)
      base.const_set('CONTENT_CLASS', base.to_s.split('::').last.downcase)
      
      def base.query(path)
        <<-QBODY
          SELECT
            #{self::ATTRIBUTE_MAPPINGS.values.map { |f| '"' + f + '"' }.join(',')}
          FROM SCOPE('shallow traversal of "#{path}"')
          WHERE "DAV:ishidden" = false
            AND "DAV:isfolder" = false
            AND "DAV:contentclass" = 'urn:content-classes:#{self::CONTENT_CLASS}'
        QBODY
      end
    end
    
    # This handy method is meant to be called from any inheriting
    # classes. It is used to bind types of folders to particular
    # Entity classes so that the folder knows what type it's
    # enumerating. So for a "calendarfolder" you'd call:
    #   set_folder_type 'calendarfolder' # or just 'calendar'
    def self.set_folder_type(dav_name)
      Folder::CONTENT_TYPES[dav_name.sub(/folder$/, '')] = self
    end
    
    # --Normally Not Used--
    # By default the CONTENT_CLASS is determined by the name
    # of your class. So for the Appointment class the
    # CONTENT_CLASS would be 'appointment'.
    # If for some reason this convention doesn't suit you,
    # you can use this method to set the appropriate value
    # (which is used in queries).
    # For example, the DAV:content-class for contacts is:
    #   'urn:content-classes:person'
    # Person doesn't strike me as the best name for our class though.
    # Most people would refer to an entry in a Contacts folder as
    # a Contact. So that's what we call our class, and we use this method
    # to make sure everything still works as it should.
    def self.set_content_class(dav_name)
      verbosity, $VERBOSE = $VERBOSE, nil # disable warnings for the next operation
      const_set('CONTENT_CLASS', dav_name)
      $VERBOSE = verbosity # revert to the original verbosity
    end
    
    # Defines what attributes are used in queries, and
    # what methods they map to in instances. You should
    # pass a Hash of method_name symbols and namespaced-attribute-name pairs.
    def self.attribute_mappings(mappings)
      
      mappings.merge! :uid => 'DAV:uid',
        :modified_at => 'DAV:getlastmodified',
        :href => 'DAV:href',
        :categories =>'http://schemas.microsoft.com/exchange/keywords-utf8'        
      
      const_set('ATTRIBUTE_MAPPINGS', mappings)

      mappings.each_pair do |k,v|
        
        define_method(k) do
          @attributes[v]
        end
        
        define_method("#{k.to_s.sub(/\?$/, '')}=") do |value|
          @attributes[v] = value
        end
        
      end
    end
    
    # Retrieve an Array of items (such as Contact, Message, etc)
    def self.find(credentials, path, conditions = nil)
      qbody = <<-QBODY
        <?xml version="1.0"?>
  			<D:searchrequest xmlns:D = "DAV:">
  				 <D:sql>
           #{conditions.nil? ? query(path) : search(path, conditions)}
           </D:sql>
        </D:searchrequest>
      QBODY
      
      response = DavSearchRequest.execute(credentials, :body => qbody)

      items = []
      xpath_query = "//a:propstat[a:status/text() = 'HTTP/1.1 200 OK']/a:prop"

      XML::Parser.string(response.body).parse.find(xpath_query).each do |node|
         items << self.new(credentials, node)        
      end
      
      items
      
    end
  end
end